helmclient.go

```diff
@@ -228,18 +228,55 @@ func (c *Client) EnsureTillerInstalled(ctx context.Context) error {
 		}
 	}
 
-	// Install the tiller deployment in the tenant cluster.
-	{
-		c.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating tiller in namespace %#q", c.tillerNamespace))
+	var err error
+	var installTiller bool
+	var pod *corev1.Pod
+	var upgradeTiller bool
 
+	{
 		o := func() error {
-			i := &installer.Options{
-				ImageSpec:      tillerImageSpec,
-				MaxHistory:     defaultMaxHistory,
-				Namespace:      c.tillerNamespace,
-				ServiceAccount: tillerPodName,
+			pod, err = getPod(c.k8sClient, tillerLabelSelector, c.tillerNamespace)
+			if IsNotFound(err) {
+				// Fall through as we need to install Tiller.
+				installTiller = true
+				return nil
+			} else if err != nil {
+				return microerror.Mask(err)
 			}
 
+			return nil
+		}
+		b := backoff.NewExponential(2*time.Minute, 5*time.Second)
+		n := backoff.NewNotifier(c.logger, context.Background())
+
+		err := backoff.RetryNotify(o, b, n)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	if pod != nil {
+		err = isTillerOutdated(pod)
+		if IsTillerOutdated(err) {
```

Comment:
For me, this is a bit confusing, maybe because of my background. But why instead of return a bool (and error if it is needed), we return an error when tiller is outdated?

Reply:
Ack. I'll add the bool param. I originally planned to use the function when the tunnel to the tiller pod is created. But during testing I found that didn't work. I agree its confusing as it is.

Reply:
With bool we have to create an error in `newTunnel`. We can also add other cases as needed. Maybe we can handle situation when installed tiller version is newer than expected. `bool` doesn't really work in `newTunnel`.

Reply:
Agreed and its nice that we have the check in `newTunnel`. As we always need a tunnel we don't have to check the tiller version in each method.